text/3804-cfg-alias.md

```diff
@@ -0,0 +1,225 @@
+
+```rust
+#![cfg_attr(some_alias, some_attribute)]
+// warning: unexpected_cfgs
+//
+// The lint could mention that `some_alias` was found in the
+// crate but is not available here.
+
+#![cfg_alias(some_alias =  true)]
+```
+
+_RFC question: "usable only after definition" is mentioned here to retain the
+ability to parse attributes in order, rather than going back and updating
+earlier attributes that may use the alias. Is this a reasonable limitation to
+keep?_
+
+_RFC question: two ways to implement this are with (1) near-literal
+substitution, or (2) checking whether the alias should be set or not at the time
+it is defined. Is there any user-visible behavior that would make us need to
+specify one or the other?_
+
+_If we go with the first option, we should limit to a single expansion to avoid
+recursing (as is done for `#define` in C)._
+
+## `cfg_alias` in non-crate attributes
+
+`cfg_alias` may also be used as a module-level attribute rather than
+crate-level:
+
+```rust
+#[cfg_alias(foo = bar)]
+mod uses_bar {
+    // Enabled/disabled based on `cfg(bar)`
+    #[cfg(foo)]
+    fn qux() { /* ... */ }
+}
+
+#[cfg_alias(foo = baz)]
+mod uses_baz {
+    // Enabled/disabled based on `cfg(baz)`
+    #[cfg(foo)]
+    fn qux() { /* ... */ }
+}
+```
+
+This has the advantage of keeping aliases in closer proximity to where they are
+used; if a configuration pattern is only used within a specific module, an alias
+can be added at the top of the file rather than making it crate-global.
+
+When defined at a module level, aliases are added to the configuration namespace
+for everything within that module including later module-level configuration.
+There is no conflict with aliases that use the same name in other modules.
+
+This RFC proposes that the use of `cfg_alias` on modules _should_ be included if
+possible. However, this may bring implementation complexity since, to the RFC
+author's knowledge, the rustc configuration system is not designed to allow
+scoped configuration. If implementation of module-level aliases turns out to be
+nontrivial, this portion of the feature may be deferred or dropped before
+stabilization.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+- This does not support more general attribute aliases, such as
+  `#![alias(foo = derive(Clone, Copy, Debug, Default)`. This seems better suited
+  for something like `declarative_attribute_macros` in [RFC3697].
+
+[RFC3697]: https://github.com/rust-lang/rfcs/pull/3697
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- The syntax `cfg_alias(name =  predicate)` was chosen to mimic assignment in
+  Rust and key-value mappings in attributes. Alternatives include:
+  - `cfg_alias(name, predicate)`, which is more similar to
+    `cfg_attr(predicate, attributes)`.
+- It may be possible to have `#[cfg_alias(...)]` work as an outer macro and only
+  apply to a specific scope. This likely is not worth the complexity.
+
+# Prior art
```

Comment:
The cfg_aliases crate could be mentioned. https://crates.io/crates/cfg_aliases

Reply:
I only have limited knowledge of the cfg_aliases crate implementation, but worked on the crate which the cfg_alias macro was based upon initially, and testing cfg_aliases it appears to suffer from similar limitations in it's cfg parsing.

In particular it has difficulty with combinations of nested and sequential vararg cfgs such as `not(any(a,b), all(c,d))`. As well as requiring build.rs.

I feel like both of these are good reasons for cargo not to defer to doing it via a crate, even if the crate is good enough to be useful for most purposes.